synapse/axon.py

```diff
@@ -1,444 +1,545 @@
 import synapse.lib.config as s_config
-import synapse.lib.msgpack as s_msgpack
 
 # for backward compat (HashSet moved from this module to synapse.lib.hashset )
 from synapse.lib.hashset import *
 
 logger = logging.getLogger(__name__)
 
-blocksize = 64 * s_const.mebibyte # 64 meg blocks
+zero64 = b'\x00\x00\x00\x00\x00\x00\x00\x00'
+blocksize = s_const.mebibyte * 64
 
-def unhex(text):
-    return binascii.unhexlify(text)
+class BlobStor(s_eventbus.EventBus):
 
-def enhex(byts):
-    return binascii.hexlify(byts).decode('utf8')
+    def __init__(self, dirn, mapsize=s_const.tebibyte):
 
-prefs = {
-    'md5': b'md5:',
-    'sha1': b'sha1:',
-    'sha256': b'sha256:',
-    'sha512': b'sha512:',
-}
+        s_eventbus.EventBus.__init__(self)
 
-class Upload:
+        path = s_common.gendir(dirn, 'blobs.lmdb')
+        self.lenv = lmdb.open(path, writemap=True, max_dbs=128)
+        self.lenv.set_mapsize(mapsize)
 
-    def __init__(self, axon, size):
+        self._blob_info = self.lenv.open_db(b'info')
+        self._blob_bytes = self.lenv.open_db(b'bytes') # <fidn><foff>=<byts>
 
-        self.gotn = 0
-        self.byts = b''
+        self._blob_clone = s_lmdb.Seqn(self.lenv, b'clone')
+        self._blob_metrics = s_lmdb.Metrics(self.lenv)
 
-        self.axon = axon
-        self.size = size
+        def fini():
+            self.lenv.sync()
+            self.lenv.close()
 
-        self.md5 = hashlib.md5()
-        self.sha1 = hashlib.sha1()
-        self.sha256 = hashlib.sha256()
-        self.sha512 = hashlib.sha512()
+        self.onfini(fini)
 
-        self.blocks = []
+    def save(self, blocs):
+        '''
+        Save items from an iterator of (<buid><indx>, <byts>).
 
-    def write(self, byts):
+        NOTE: This API is only for use by a single Axon who owns this BlobStor.
+        '''
+        with self.lenv.begin(write=True, db=self._blob_bytes) as xact:
 
-        blen = len(byts)
+            size = 0
+            count = 0
+            clones = []
 
-        self.gotn += blen
+            for lkey, lval in blocs:
 
-        if self.gotn > self.size:
-            raise Exception('upload larger than expected')
+                size = len(lval)
+                clones.append(lkey)
 
-        self.md5.update(byts)
-        self.sha1.update(byts)
-        self.sha256.update(byts)
-        self.sha512.update(byts)
+                xact.put(lkey, lval, db=self._blob_bytes)
 
-        self.byts += byts
+                size += len(lval)
+                count += 1
 
-        if self.gotn < self.size and len(self.byts) < blocksize:
-            return False
+            self._blob_clone.save(xact, clones)
+            self._blob_metrics.inc(xact, 'bytes', size)
+            self._blob_metrics.inc(xact, 'blocks', count)
 
-        with self.axon._withLmdbXact(write=True) as xact:
+            tick = s_common.now()
+            self._blob_metrics.record(xact, {'time': tick, 'size': size, 'blocks': count})
```

Comment:
would having 'took' here be reasonable? similar to cyrotanks?

Reply:
yes this is gonna be plumbed